File: app/src/main/java/com/github/gotify/messages/provider/ApplicationHolder.java
package com.github.gotify.messages.provider;

import android.app.Activity;
import com.github.gotify.Utils;
import com.github.gotify.api.ApiException;
import com.github.gotify.api.Callback;
import com.github.gotify.client.ApiClient;
import com.github.gotify.client.api.ApplicationApi;
import com.github.gotify.client.model.Application;
import java.util.Collections;
import java.util.List;

public class ApplicationHolder {
    private List<Application> state;
    private Runnable onUpdate;
    private Runnable onUpdateFailed;
    private Activity activity;
    private ApiClient client;

    public ApplicationHolder(Activity activity, ApiClient client) {
        this.activity = activity;
        this.client = client;
    }

    public void requestIfMissing() {
        if (state == null) {
            request();
        }
    }

    public void request() {
        client.createService(ApplicationApi.class)
                .getApps()
                .enqueue(Callback.callInUI(activity, this::onReceiveApps, this::onFailedApps));
    }

    private void onReceiveApps(List<Application> apps) {
        state = apps;
        if (onUpdate != null) onUpdate.run();
    }

    private void onFailedApps(ApiException e) {
        Utils.showSnackBar(activity, "Could not request applications, see logs.");
        if (onUpdateFailed != null) onUpdateFailed.run();
    }

    public List<Application> get() {
        return state == null ? Collections.emptyList() : state;
    }

    public void onUpdate(Runnable runnable) {
        this.onUpdate = runnable;
    }

    public void onUpdateFailed(Runnable runnable) {
        this.onUpdateFailed = runnable;
    }
}

File: app/src/main/kotlin/com/github/gotify/messages/provider/ApplicationHolder.kt
package com.github.gotify.messages.provider

import android.app.Activity
import com.github.gotify.Utils
import com.github.gotify.api.Callback
import com.github.gotify.client.ApiClient
import com.github.gotify.client.api.ApplicationApi
import com.github.gotify.client.model.Application

internal class ApplicationHolder(private val activity: Activity, private val client: ApiClient) {
    private var state = listOf<Application>()
    private var onUpdate: Runnable? = null
    private var onUpdateFailed: Runnable? = null

    fun wasRequested() = state.isNotEmpty()

    fun request() {
        client.createService(ApplicationApi::class.java)
            .apps
            .enqueue(
                Callback.callInUI(
                    activity,
                    onSuccess = Callback.SuccessBody { apps -> onReceiveApps(apps) },
                    onError = { onFailedApps() }
                )
            )
    }

    private fun onReceiveApps(apps: List<Application>) {
        state = apps
        if (onUpdate != null) onUpdate!!.run()
    }

    private fun onFailedApps() {
        Utils.showSnackBar(activity, "Could not request applications, see logs.")
        if (onUpdateFailed != null) onUpdateFailed!!.run()
    }

    fun get() = state

    fun onUpdate(runnable: Runnable?) {
        onUpdate = runnable
    }

    fun onUpdateFailed(runnable: Runnable?) {
        onUpdateFailed = runnable
    }
}
